src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1291 @@
+
+				rc = holtek8_macro_to_events(button, macro_events);
+				if (rc < 0)
+					return rc;
+
+				rc = holtek8_write_macro_data(device, macro_events);
+				if (rc < 0)
+					return rc;
+
+				data->type = HOLTEK8_BUTTON_TYPE_MACRO;
+				data->macro.mode = HOLTEK8_BUTTON_MACRO_REPEAT_COUNT;
+				data->macro.index = rc;
+
+				return 0;
+			}
+			if (rc < 0) {
+				return rc;
+			}
+
+			data->type = HOLTEK8_BUTTON_TYPE_KEYBOARD;
+			data->keyboard.modifiers = modifiers;
+			data->keyboard.hid_key = ratbag_hidraw_get_keyboard_usage_from_keycode(device, key1);
+			data->keyboard.hid_key2 = ratbag_hidraw_get_keyboard_usage_from_keycode(device, key2);
+
+			if (rc == 1 && modifiers == 0 && data->keyboard.hid_key == 0) {
+				uint16_t hid_code_cc = ratbag_hidraw_get_consumer_usage_from_keycode(device, key1);
+				if (hid_code_cc == 0)
+					break;
+
+				data->type = HOLTEK8_BUTTON_TYPE_MEDIA;
+				set_unaligned_le_u16(data->media.hid_key, hid_code_cc);
+			}
+
+			break;
+		}
+		default:
+			log_error(device->ratbag, "Button %u action type unsupported: %u\n", button->index, button->action.type);
+			return -EINVAL;
+	}
+
+	return 0;
+}
+
+const struct holtek8_sensor_config *
+holtek8_get_sensor_config(enum holtek8_sensor sensor)
+{
+	const struct holtek8_sensor_config *cfg = NULL;
+
+	ARRAY_FOR_EACH(holtek8_sensor_configurations, cfg) {
+		if (cfg->sensor != sensor) {
+			continue;
+		}
+		return cfg;
+	}
+
+	abort();
+}
+
+enum holtek8_sensor
+holtek8_get_sensor_from_name(const char *name)
+{
+	const struct holtek8_sensor_config *cfg = NULL;
+
+	ARRAY_FOR_EACH(holtek8_sensor_configurations, cfg) {
+		if (!streq(cfg->name, name)) {
+			continue;
+		}
+		return cfg->sensor;
+	}
+
+	return HOLTEK8_SENSOR_UNKNOWN;
+}
+
+/*
+ * The device is sensitive to unsynchronized writes.
+ * This functions asks the device if it really expects the amount
+ * of bytes to write we think.
+ *
+ * @return 0 on success or a negative errno.
+ */
+static int
+holtek8_poll_write_ready(struct ratbag_device *device, uint8_t bytes_left)
+{
+	struct holtek8_data *drv_data = device->drv_data;
+	struct holtek8_feature_report report;
+	uint8_t bytes_left_dev, bytes_left_pos;
+	int i, rc;
+
+	switch (drv_data->api_version) {
+		case HOLTEK8_API_A:
+			bytes_left_pos = 3;
+			break;
+		case HOLTEK8_API_B:
+			bytes_left_pos = 1;
+			break;
+	}
+
+	for (i = 0; i < HOLTEK8_POLL_RETRY_LIMIT; i++) {
+		rc = holtek8_get_feature_report(device, &report);
+		if (rc < 0)
+			return rc;
+
+		bytes_left_dev = report.arg[bytes_left_pos];
+
+		if (bytes_left == bytes_left_dev)
+			return 0;
+
+		msleep(HOLTEK8_POLL_TIME_MS);
+	}
+
+	return -EIO;
+}
+
+/*
+ * Clears hidraw's read buffer.
+ * Prevents reading incorrect data if there was another
+ * configuration program before us.
+ *
+ * @return Number of chunks of data cleared or a negative errno.
+ */
+static int
+holtek8_clear_read_buffer(struct ratbag_device *device)
```

Comment:
Not a blocker, but if this is needed, it should probably be a helper in libratbag namespace instead.